Add docs recommending autoscaling setup (#324)
temporalio/temporal-worker-controller: docs/scaling-recommendations.md

> # Scaling Recommendations: HPA + prometheus-adapter vs KEDA

Reviewer: ```suggestion
# Scaling Recommendations
```

> This document describes practical reactivity and reliability tradeoffs when scaling Temporal workers per worker-deployment-version on Kubernetes, and recommends which tool fits which workload pattern.

Reviewer: ```suggestion
This document describes practical reactivity and reliability tradeoffs when scaling Temporal workers per worker deployment version on Kubernetes, and recommends which tool fits which workload pattern.
```

here and throughout the doc... we use "worker deployment version", not "worker-deployment-version", elsewhere in the docs.

The `internal/demo/` example wires the HPA path described here. The KEDA path is mentioned for comparison and as a recommendation for workloads that cannot tolerate the HPA path's limits.

> ## TL;DR — Pick by workload pattern

Reviewer: ```suggestion
## TL;DR

We recommend choosing a scaler approach that aligns with the workload pattern your application exhibits.
```

| Workload pattern | Recommendation |
|------------------|----------------|

> | Continuous traffic (task queue always loaded) | HPA + prometheus-adapter, scaling on slot utilization + backlog count |

Reviewer: ```suggestion
| Continuous traffic (task queue always loaded) | HPA |
```

> | Idle periods >5 min between work; needs scale-from-zero | KEDA Temporal scaler |

Reviewer: ```suggestion
| Idle periods >5 min between work OR needs scale-from-zero | KEDA Temporal scaler |
```

| Required reactivity < ~60 s from first backlog | KEDA Temporal scaler |

> | Required reactivity ~90 s typical, tolerant of occasional multi-minute stalls | HPA + prometheus-adapter is fine |

Reviewer: ```suggestion
| Required reactivity ~90 s typical, tolerant of occasional multi-minute stalls | HPA + prometheus-adapter |
```

> | Very large namespaces (N × M HPAs polling) | HPA + prometheus-adapter; KEDA hits the 50 RPS namespace rate limit |

Reviewer: ```suggestion
| 1000s of task queues and worker deployment versions  | HPA + prometheus-adapter |
```

Reviewer 2: I would recommend clarifying that this is talking about Temporal Cloud...OSS customers won't have this limitation OOTB. 
I would also recommend referencing https://docs.temporal.io/cloud/limits and if this limit is not explicitly there have it added there. Ideally, you'd have a special `Worker Versioning` section on that limits page so customers can just reference that directly

Author: OSS customers can increase the rate limit, but they still will experience the default rate limit unless they override it. If they do increase the rate limit, they will also need to eventually scale up the per-namespace worker, because the per-namespace worker handles `DescribeWorkerDeployment` requests.

> ## The reactivity model for HPA + prometheus-adapter

Reviewer: ```suggestion
## HPA scaling signal

This section describes the signal used by HPA + prometheus adapter to adjust the count of workers in a Kubernetes deployment managed by Temporal Worker Controller.

There are two metric data points that are scraped by HPA + prometheus adapter.

`temporal_cloud_v1_approximate_backlog_count` (or just "backlog") is a measurement of the number of pending tasks on a particular task queue that are waiting for a poller (a worker) to pull that task and process it.

`temporal_slot_utilization` (or just "slot util") is emitted directly by worker pods (no Temporal Cloud aggregation), scraped at the ServiceMonitor interval (~10–30 s), and reflects the current state of a particular worker. This metrics rises *before* backlog accumulates — slots saturate first, then queueing starts.
```

For a continuously-loaded task queue, the end-to-end delay from "backlog appears" to "HPA scales up" decomposes as:

```
backlog appears at T0
  └─ Temporal Cloud OpenMetrics emission cadence    +~60 s worst-case  (~1 sample/minute)
       └─ Prometheus scrape interval                 +~10 s
            └─ HPA poll interval                     +~15 s
                 └─ scale-up stabilization window    +~your config
                      └─ first replica added
```

> **Typical end-to-end reactivity is ≈ 85 seconds + your stabilization window.** Empirically, sample age in Prometheus for a single series follows a sawtooth between 0 and 60 seconds (matching the gateway's ~1/min emission cadence). p50 sample age ≈ 30s, p95 ≈ 50s. The 60-second emission cadence is the inherent floor — smaller scrape intervals, tighter `metricsRelistInterval`, or recording rules cannot improve it because they all consume the same upstream cadence.

Reviewer: What is "your stabilization window"? That is not defined anywhere and this reads like Claude has seen the phrase "stabilization window" in other docs that it scraped from the Internet that described auto-scaling algorithms but doesn't actually understand what "stabilization window" means.

Also, "Typical end-to-end reactivity" doesn't make sense here and sounds like a term Claude either made up or has hallucinated-adopted from the term "end-to-end reactivity" [from frontend software development patterns](https://www.explainthis.io/en/swe/understanding-reactivity-from-frontend).

Author: Sorry about any Claude-isms, the reason this is in Draft mode is because I hadn't done a full pass over it yet. I care deeply about not putting PRs in front of people that I haven't reviewed myself and don't necessarily endorse any of this until it's out of draft mode.

Claude probably came up with this after seeing a bunch of grafana screenshots that I sent it.

### Caveat: gateway delivery delay

> During our investigation we observed periods of several minutes during which Temporal Cloud's OpenMetrics endpoint returned the same embedded timestamps on repeated scrapes for *every* series across the account simultaneously — backlog series, action counts, error counts, every queue, every namespace, all showing identical staleness to the second (e.g. all ~30 visible series reading 239s old at once). The Prometheus scrape continued to succeed (`up{job="temporal_cloud"}` stayed 1, HTTP 200 responses) — the response body simply repeated already-known samples instead of advancing.

Reviewer 3: this reads odd given that the project is from the same people as Temporal Cloud - can this be reworded?

Once the delay resolved, the gateway delivered the missing samples with their original minute-aligned timestamps in a burst, so Prometheus's storage ends up with a complete 1/minute series in retrospect. We verified this directly: across a 3-hour window covering one such delay event, every gap between consecutive sample timestamps was exactly 60 seconds, no exceptions.

The retrospective completeness is helpful for dashboards and post-hoc analysis, but it does **not** help an HPA, which queries the *latest available* value at decision time. During a delivery delay, the latest available sample is the one from before the delay started. The HPA sees real staleness even though the underlying record will eventually be filled in.

> We have only directly characterized this pattern during one investigation session (seeing it twice in ~2 hours of close observation). Frequency in normal operation is not yet known and is open with Temporal's Observability team. If your workload cannot tolerate occasional multi-minute scaling pauses, prefer KEDA.

Reviewer 3: this sounds like sharing internal sausage making - which as a customer I am not sure what to take away from

Reviewer: Right, which is why I recommended in my review that this entire section be removed :)

> This is also why `metricsRelistInterval: 5m` is the recommended setting: the discovery window must comfortably exceed the longest expected delay so the metric does not deregister, otherwise re-registration waits up to one more relist cycle after delivery resumes.

Reviewer: Recommend removing this section. I've moved some of the content and reworded it in the `## HPA limitations` section below.

### Slot utilization is a much faster leading signal

`temporal_slot_utilization` is emitted directly by worker pods (no Temporal Cloud aggregation), scraped at the ServiceMonitor interval (~10–30 s), and reflects current state. It also rises *before* backlog accumulates — slots saturate first, then queueing starts. So a two-metric HPA with both slot util and backlog gives you fast scale-up via slot util and a backlog-driven backstop.

> The demo HPA uses both. For production scaling we recommend keeping both as well.

Reviewer: Recommend removing this. I've pulled some of the content into the proposed new `## HPA strengths` and `## HPA scaling signal` sections

### Mode 2: series-level silence (common in low-traffic workloads)
- Trigger: a task queue with no polls or new tasks for >5 minutes. Temporal unloads it from memory and stops emitting `temporal_cloud_v1_approximate_backlog_count` for that specific `(task_queue, build_id, ...)` labelset. Other queues' series continue to emit.
- Symptom in HPA events: `no metrics returned from external metrics API`. The metric *name* is still registered; the HPA's specific label selector just matches zero rows now.
- Recovery: traffic resumes → queue reloads → next emission cycle (~1 min) + 3-min aggregation lag → HPA can read value again.

In a two-metric HPA configured with slot utilization, this is mostly fine: the HPA reports `ScalingActive=True` based on slot utilization while backlog is unavailable, and rejoins backlog scaling once it returns. We've confirmed this empirically in this demo cluster — the HPA continued scaling correctly on slot utilization through 1000+ backlog `FailedGetExternalMetric` events.

> ## Why this demo does not use a backlog recording rule

Reviewer 3: which demo? "this" is an unclear reference

Reviewer: Ii recommended removing this entire section

> A prior version of this demo wrapped the raw Cloud series in a Prometheus recording rule:

Reviewer 3: ```suggestion
A prior version of this demo wrapped the raw Temporal Cloud series in a Prometheus recording rule:
```

```yaml
- record: temporal_approximate_backlog_count
  expr: sum by (...) (temporal_cloud_v1_approximate_backlog_count)
```

The rule was originally added to work around a label-formatting issue in an older Temporal Cloud release. With native per-version labels (`temporal_worker_deployment_name`, `temporal_worker_build_id`) now opt-in, the rule no longer earns its keep:

- **It doesn't reduce reactivity.** The HPA reactivity floor is the upstream OpenMetrics emission cadence (~60s), not anything the rule could fix.
- **It duplicates the cardinality bill.** Per-`(task_queue, build_id)` labels are already opt-in at the OpenMetrics level *because* of cardinality. Adding a recording rule on top means storing the same high-cardinality series twice.
- **It hides a `sum(...)` that the adapter already does.** prometheus-adapter's `metricsQuery: sum(<<.Series>>{<<.LabelMatchers>>})` performs the same collapsing at query time. Pedagogically, "the adapter does the sum" is cleaner than "a recording rule sums first, then the adapter sums again."

> - **It does not solve series-level silence.** When the source goes silent (task queue unloaded), the rule output also goes silent eventually (once Prometheus's staleness lookback expires).

Reviewer: Remove this. It's not necessary and likely will just confuse the reader.

What the recording rule *does* buy is registration stability after operational events: when the source series is sparse-by-timestamp, the rule produces a dense 10-second sample stream that lets the adapter discover with a tight `metricsRelistInterval`. If you find yourself fighting registration flicker on every adapter restart and would rather pay the cardinality cost than tune `metricsRelistInterval`, a recording rule is a reasonable choice. Otherwise, prefer the raw metric.

In this demo we set `metricsRelistInterval: 5m` and consume the raw metric directly.

> 

Reviewer: ```suggestion
## HPA strengths

Because HPA uses a single OpenMetrics scrape to gather all series for the namespace in a single HTTP request, the HPA approach scales independently of namespace count. The single HTTP request for OpenMetrics more efficient than KEDA's Temporal API-based approach, and will not run into Temporal API rate limiting problems (see section below on [KEDA limitations](#keda-limitations)).

HPA + prometheus adapter configured to look at both slot util and backlog provides fast scale-up via slot util and a backlog-driven backstop to prevent overly reactive replica count adjustment.
```

> ## Why prometheus-adapter cannot do scale-from-zero

Reviewer: ```suggestion
## HPA limitations

This section describes two known limitations for HPA + prometheus adapter.

Temporal Cloud's OpenMetrics endpoint may sometimes return the same embedded timestamps on repeated scrapes for each series across the account simultaneously — backlog series, action counts, error counts, every queue, every namespace. This delay in returning fresh metrics data can impact the speed to which HPA + prometheus adapter scales out or in the replica count for a worker deployment version. This means that HPA + prometheus adapter may not be a good solution if your workload cannot tolerate occasional multi-minute scaling pauses.

> **Note**: This is why `metricsRelistInterval: 5m` is the recommended setting: the discovery window must comfortably exceed the longest expected delay so the metric does not deregister, otherwise re-registration waits up to one more relist cycle after delivery resumes.
```

> Scale-from-zero on backlog through the metric path requires the metric to exist while there are zero workers. It does not:

Reviewer: ```suggestion
HPA cannot scale your Worker Deployment from zero because the signal for scaling does not yet exist. The signal for scaling is the backlog metric for the task queue associated with the workers in the Worker Deployment. This metric will not exist until there is at least one worker polling the task queue.
```

1. Zero workers means no polls.
2. No polls for >5 minutes means the task queue is unloaded from Temporal Cloud's memory.
3. An unloaded queue emits no metric.
4. Adapter discovery returns no series, or HPA queries return no rows.

> 5. HPA cannot scale up because there's no signal to scale on.

Reviewer: ```suggestion
In addition to the "first worker start" problem, for customers using Temporal Cloud, if there are no polling workers for a task queue for more than 5 minutes, Temporal Cloud will unload the task queue from memory. Unloaded task queues do not emit metrics, and therefore the signal that HPA uses to scale up will not be present.
```

Reviewer 3: Isn't the same problem also seen with OSS?

Reviewer: not sure. @carlydf do you know the answer to this?

Author: yes, same problem for OSS and Cloud here, Claude hallucinating. (PR in draft mode because I had not fully reviewed yet, didn't intend to waste your energy on review until it was ready!)

> Submitting a workflow does load the task queue back into memory, but the metric still won't reach the HPA until the next OpenMetrics emission cycle (~1 minute) plus scrape and HPA poll. By the time the HPA reacts, you've already had ~1+ minute of unprovisioned work.

Reviewer: ```suggestion
Submitting a workflow does load the task queue back into memory, but the metric still won't reach the HPA until the next OpenMetrics emission cycle (~1 minute). By the time the HPA reacts, you've already had ~1+ minute of unprovisioned work.
```

> KEDA's Temporal scaler calls `DescribeTaskQueue(stats=true)` (or `DescribeWorkerDeploymentVersion`), which loads the queue synchronously and returns the backlog directly. No metric pipeline involved. Scale-from-zero in seconds.

Reviewer: ```suggestion
## KEDA strengths

KEDA's Temporal scaler calls `DescribeTaskQueue(stats=true)` (or `DescribeWorkerDeploymentVersion`), which loads the queue synchronously and returns the backlog directly. This allows KEDA to scale Temporal workers from zero.
```

> ## When KEDA hits its own limits

Reviewer: ```suggestion
## KEDA limitations
```

> KEDA bypasses the metric pipeline but uses Temporal API calls, which are subject to a per-namespace rate limit:

Author: Other KEDA limitation (for now) https://github.com/temporalio/temporal-worker-controller/issues/355

KEDA also does not actually work with TWC until https://github.com/temporalio/temporal-worker-controller/issues/286 is closed. Luckily we have an open community PR #351 that will add support for the KEDA temporal trigger, which I think can be merged soon. I just reviewed it.

Reviewer: Yup. Noted. https://github.com/temporalio/temporal-worker-controller/pull/351 looks almost ready to merge.

```
FrontendGlobalWorkerDeploymentReadRPS = 50  # per namespace, evenly distributed across frontend instances
```

For a namespace with N task queues × M worker-deployment-versions = K HPAs, each KEDA poll uses ~1 API call. The polling budget:

| HPA count | Poll every 30s | Poll every 10s | Poll every 5s |
|-----------|----------------|----------------|---------------|
| 50        | 1.7 RPS (3%)   | 5 RPS (10%)    | 10 RPS (20%)  |
| 250       | 8 RPS (17%)    | 25 RPS (50%)   | 50 RPS (100%) |
| 1500      | 50 RPS (100%)  | exceeds limit  | exceeds limit |

> prometheus-adapter has no equivalent per-namespace bottleneck — one OpenMetrics scrape returns all series for the namespace in a single HTTP request, scaling independently of HPA count.

Reviewer: ```suggestion
```

(this section is about KEDA's limits, not HPA's. We already cover HPA's limitations in the previous section. I've moved this to a "HPA strengths" section above.)

> So for very large namespaces (hundreds–thousands of HPAs) needing fast reactivity, neither path is great: KEDA hits the API rate limit, and the metric path has the 3-min aggregation floor. In practice this is a "talk to your account team" situation.

Reviewer: ```suggestion
If you are using KEDA with Temporal Cloud and hitting the API rate limit described above, you will need to contact your Temporal Cloud account team to discuss increasing the rate limits.
```

For a continuously-loaded task queue, important events from "backlog appears" to "HPA scales up" can be visualized like so:

```
backlog appears at T0
  └─ Temporal Cloud OpenMetrics emission cadence     + ~60s worst-case  (~1 sample/minute)
       └─ Prometheus scrape interval                 + ~10s
            └─ HPA poll interval                     + ~15s
                 └─ scale-up stabilization window    + taken from HPA configuration
                      └─ first replica added
```
[tc-openmetrics]: https://docs.temporal.io/cloud/metrics/openmetrics

## HPA strengths

Because HPA uses a single OpenMetrics scrape to gather all series for the namespace in a single HTTP request, the HPA approach scales independently of namespace count. The single HTTP request for OpenMetrics more efficient than KEDA's Temporal API-based approach, and will not run into Temporal API rate limiting problems (see section below on [KEDA limitations](#keda-limitations)).

HPA + prometheus adapter configured to look at both slot util and backlog provides fast scale-up via slot util and a backlog-driven backstop to prevent overly reactive replica count adjustment.

> ## HPA limitations

Reviewer 4: this is great! was wondering - should we also mention that our OM endpoint inherently has a 3 minute time lag, which would mean that at time=0, we are seeing a one minute aggregate of time.now() - 3 minutes

Reviewer: @Shivs11 I think you meant "initially" instead of "inherently"? It's true that there is a 3-minute potential *initial lag* ([ref](https://docs.temporal.io/cloud/metrics/openmetrics#overview)) but this is only relevant for brand new task queues or task queues that have been deregistered from the OM endpoint. This is why there is a note in the doc:

> **Note**: This is why `metricsRelistInterval: 5m` is the recommended setting: the discovery window must comfortably exceed the longest expected delay so the metric does not deregister, otherwise re-registration waits up to one more relist cycle after delivery resumes.

I've added a warning about the initial 3 minute potential export lag.

This section describes two known limitations for HPA + prometheus adapter.

Temporal Cloud's OpenMetrics endpoint may sometimes return the same embedded timestamps on repeated scrapes for each series across the account simultaneously — backlog series, action counts, error counts, every queue, every namespace. This delay in returning fresh metrics data can impact the speed to which HPA + prometheus adapter scales out or in the replica count for a worker deployment version. This means that HPA + prometheus adapter may not be a good solution if your workload cannot tolerate occasional multi-minute scaling pauses.

> **Note**: This is why `metricsRelistInterval: 5m` is the recommended setting: the discovery window must comfortably exceed the longest expected delay so the metric does not deregister, otherwise re-registration waits up to one more relist cycle after delivery resumes.

> HPA cannot scale your Worker Deployment from zero because the signal for scaling does not yet exist. The signal for scaling is the backlog metric for the task queue associated with the workers in the Worker Deployment. This metric will not exist until there is at least one worker polling the task queue.

Reviewer 4: i took about a second to understand what this really meant - at first, I thought this meant that there won't be a backlog metric emitted if you don't have workers running at all (which is not true since you do have this metric being emitted for the unversioned world without workers being present)

I know you have clearly mentioned versions in the preamble here, but do you think we can be extra clear and mention the backlog count per version is not emitted without a worker being present since that is what creates a version in temporal?

Reviewer: @Shivs11 do we really need to care about the unversioned world in TWC? I mean, TWC doesn't deal with anything that isn't versioned since it automatically creates WorkerDeploymentVersions for new worker image tags...

Author: FWIW, I think the "first ever" scale-from-zero is almost not worth mentioning in TWC-land, because the TWC always creates a new worker version with at least 1 replica, thereby creating the queue and metric for that version. I think it could be better to just focus on the queue idleness problem (which I feel you've explained well).

## TL;DR

We recommend choosing a scaler approach that aligns with the workload pattern your application exhibits.

| Workload pattern | Recommendation |
|------------------|----------------|
| Continuous traffic (task queue always loaded) | HPA |
| Idle periods >5 min between work OR needs scale-from-zero | KEDA Temporal scaler |
| Required reactivity < ~60 s from first backlog | KEDA Temporal scaler |
| Required reactivity ~90 s typical, tolerant of occasional multi-minute stalls | HPA + prometheus-adapter |
| 1000s of task queues and worker deployment versions  | HPA + prometheus-adapter |

## HPA scaling signal

This section describes the signal used by HPA + prometheus adapter to adjust the count of workers in a Kubernetes deployment managed by Temporal Worker Controller.

There are two metric data points that are scraped by HPA + prometheus adapter.

`temporal_cloud_v1_approximate_backlog_count` (or just "backlog") is a measurement of the number of pending tasks on a particular task queue that are waiting for a poller (a worker) to pull that task and process it. This is a metric provided by [Temporal Cloud's OpenMetrics aggregation service][tc-openmetrics].

> `temporal_slot_utilization` (or just "slot util") is emitted directly by Workers (no Temporal Cloud aggregation), scraped at the Prometheus `ServiceMonitor` interval (~10–30 s), and reflects the current state of a particular Worker. This metric rises *before* backlog accumulates. In other words, slots on the Worker saturate first, then queueing starts.

Reviewer 5: Is there a documentation about worker slot setup? In our Datadog environment we ended up setting the default for slot utilization to `1000` and we noticed that we never actually get anywhere close in using them up. It might be worth mentioning for the customers that if this is not properly adjusted for the worker's resources the scaling might not work as expected.

Reviewer: @eniko-dif I've added a link to this doc: https://docs.temporal.io/develop/worker-performance

Unfortunately, I don't think there's going to be a one-size-fits-all recommendation for slot utilization values because this is going to be dependent on the rate of the user's workflows (and their composing activities) being executed. Obviously that's going to be highly user-specific. I've added a link to the docs about [choosing an appropriate slot supplier](https://docs.temporal.io/develop/worker-performance#choosing-slot-supplier-types) and highlighted this callout from that doc:

> Scenarios with tasks that have variable, or very high, per-task resource needs should rely on fixed-size suppliers and manual tuning rather than resource-based suppliers.

For a continuously-loaded task queue, important events from "backlog appears" to "HPA scales up" can be visualized like so:

```
backlog appears at T0
  └─ Temporal Cloud OpenMetrics emission cadence     + ~60s worst-case  (~1 sample/minute)
       └─ Prometheus scrape interval                 + ~10s
            └─ HPA poll interval                     + ~15s
                 └─ scale-up stabilization window    + taken from HPA configuration
                      └─ first replica added
```
[tc-openmetrics]: https://docs.temporal.io/cloud/metrics/openmetrics

> 

Reviewer 5: nit: i'd mention here that an example configuration is available a bit below (maybe with a link to jump through keda if the user is not interested)

Reviewer: I moved the recommended config up.

```yaml
metricsRelistInterval: 5m   # must accommodate Cloud's ~3-min embedded-timestamp lag
rules:
  external:
    - seriesQuery: 'temporal_cloud_v1_approximate_backlog_count{temporal_worker_build_id!="__unversioned__"}'
      metricsQuery: 'sum(<<.Series>>{<<.LabelMatchers>>})'
      name:
        as: "temporal_cloud_v1_approximate_backlog_count"
      resources:
        namespaced: false
```

The `seriesQuery` filter excludes `__unversioned__` series. Without it, accounts with many unversioned namespaces produce 5000+ series in the discovery response, which slows or breaks adapter discovery. The filter scopes discovery to versioned workloads.

**HPA template** (`examples/wrt-hpa-backlog.yaml`): two metrics — slot utilization (fast leading signal, scale-up gate) and backlog count (confirming signal, AverageValue target).

```yaml
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
spec:
  scaleTargetRef: {}
  minReplicas: 1
  maxReplicas: 30
  metrics:
    - type: External
      external:
        metric:
          name: temporal_slot_utilization
          selector:
            matchLabels:
              worker_type: "ActivityWorker"
        target:

>           type: Value

Reviewer 5: nit: I'd add the comment from the `wrt-hpa-backlog` example why this is a Value, because upon first look I was confused.

Reviewer: I took this directly from the existing examples in the demo/ directory :) But I agree with you that it should more correctly be `AverageValue` and a value of `"0.75"` not `"750m"`. Updated.

          value: "750m"

    - type: External
      external:
        metric:
          name: temporal_cloud_v1_approximate_backlog_count
          selector:
            matchLabels:
              temporal_task_queue: "default_helloworld"
              task_type: "Activity"
        target:
          type: AverageValue

>           averageValue: "1"

Reviewer 5: Isn't 1 a bit too low?

Reviewer: Not necessarily... depends on the workload... for a high-request-rate workflow/taskqueue, with workers struggling to process the incoming requests, the backlog count might be much higher than 1, but it all depends on the workload...

>   behavior:

Reviewer 5: Isn't it an issue that HPA would scale up if the system is well-sized? (meaning that the backlog count is not building up, but the workers have a high slot utilization)

Reviewer: If this happens, the user would want to adjust the AverageValue target value for temporal_slot_utilization so that it would not trigger a scale up, no? Or adjust the stabilization window...

Author: I think a lot of use cases target zero backlog, so scaling up as soon as any backlog is detected makes sense (especially given detection delays)

    scaleUp:
      stabilizationWindowSeconds: 30
      policies:

>         - type: Percent

Reviewer 5: I'd personally put quicker scale up values than scale down.

Reviewer: Isn't that what this example shows?

# Scaling Recommendations

This document describes practical reactivity and reliability tradeoffs when scaling Temporal workers per worker deployment version on Kubernetes, and recommends which tool fits which workload pattern.

The `internal/demo/` example wires the HPA path described here. The KEDA path is mentioned for comparison and as a recommendation for workloads that cannot tolerate the HPA path's limits.

## TL;DR

We recommend choosing a scaler approach that aligns with the workload pattern your application exhibits.

| Workload pattern | Recommendation |
|------------------|----------------|

> | Continuous traffic (task queue always loaded) | HPA |

Author: nit: is there a difference between `HPA + prometheus adapter` in lower rows and just `HPA` here? Can we just say `HPA + Temporal Cloud Metrics + metrics adapter` in all the HPA rows? For example, Datadog metrics can be passed to HPA with per-version labels as described [here](https://docs.datadoghq.com/containers/guide/cluster_agent_autoscaling_metrics/?tab=datadogoperator#autoscaling-without-datadogmetric-queries). So someone could use the Temporal OpenMetrics endpoint to get backlog count into their Datadog system, and then use Datadog HPA adapter to use HPA + Datadog metrics. We just have not worked that example because Datadog is paid only.

Author: Or, maybe we say **HPA + prometheus adapter*** in all those rows, and let the * lead to the paragraph I propose in my other comment: "we tested and are discussing prometheus in detail, but there are other ways to pipe Cloud Metrics -> HPA that have similar caveats with slightly different timing depending on configuration"

Reviewer: Updated all lines to "HPA + prometheus-adapter" and added a note with your text.

> | Idle periods >5 min between work OR needs scale-from-zero | KEDA Temporal scaler |

Author: I think this is AND and not OR.

If there are idle periods >5 min between work, but pod count is >=1, then the pollers in the one pod will keep the task queues loaded.

Basically:

Task queue always loaded (continuous task traffic OR never scale-from-zero) --> HPA

Task queue may be unloaded (idle periods >5 min between tasks AND needs scale-from-zero) --> KEDA Temporal scaler

Reviewer: ack. changed to AND.

| Required reactivity < ~60 s from first backlog | KEDA Temporal scaler |
| Required reactivity ~90 s typical, tolerant of occasional multi-minute stalls | HPA + prometheus-adapter |
| 1000s of task queues and worker deployment versions  | HPA + prometheus-adapter |

## HPA scaling signal

> This section describes the signal used by HPA + prometheus adapter to adjust the count of workers in a Kubernetes deployment managed by Temporal Worker Controller.

Author: Maybe here is a place to say

> HPA + metrics adapter.... We will discuss the Prometheus Metrics Adapter in depth because it is free to install so we have tested it with the Worker Controller end to end, but the Temporal Cloud Metrics -> HPA method is expected to work with other metrics providers as long as (1) the metrics provider can ingest Temporal Cloud Metrics from our [OpenMetrics endpoint](https://docs.temporal.io/cloud/metrics/openmetrics), and (2) the metrics provider has a Kubernetes adapter to pipe those metrics to HPA and use native HPA `matchLabels` format to create a per-version metrics query. Using a different aggregation layer could incur additional delays, depending on the configuration of that layer. We will discuss the Prometheus pathway in detail.
>
> For example:
> - Prometheus
>   - [Temporal Cloud Metrics integration](https://docs.temporal.io/cloud/metrics/openmetrics/metrics-integrations#prometheus-grafana) (Temporal Cloud -> Prometheus)
>   - [HPA adapter](https://github.com/kubernetes-sigs/prometheus-adapter) (Prometheus -> HPA)
>
> - Datadog
>   - [Temporal Cloud Metrics integration](https://docs.temporal.io/cloud/metrics/openmetrics/metrics-integrations#datadog) (Temporal Cloud -> DataDog)
>   - [HPA adapter](https://docs.datadoghq.com/containers/guide/cluster_agent_autoscaling_metrics/?tab=datadogoperator#autoscaling-without-datadogmetric-queries) (DataDog -> HPA)
>
> - New Relic
>   - [Temporal Cloud Metrics integration](https://docs.temporal.io/cloud/metrics/openmetrics/metrics-integrations#new-relic) (Temporal Cloud -> New Relic)
>   - [HPA adapter](https://docs.newrelic.com/docs/kubernetes-pixie/kubernetes-integration/advanced-configuration/newrelic-metrics-adapter/) (New Relic -> HPA)
>
> - OpenTelemetry Collector -> system of your choosing
>   - [Temporal Cloud Metrics integration](https://docs.temporal.io/cloud/metrics/openmetrics/metrics-integrations#opentelemetry-collector-configuration) (Temporal Cloud -> OpenTelemetry)
>   - [AWS CloudWatch OpenTelemetry integration](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-OpenTelemetry-metrics.html) (OpenTelemetry -> CloudWatch)
>   - [HPA adapter for AWS CloudWatch](https://aws.amazon.com/blogs/compute/scaling-kubernetes-deployments-with-amazon-cloudwatch-metrics/) (CloudWatch -> HPA)

Reviewer: Added with slight rewording.

                      └─ first replica added
```

Per-Worker tunable configuration options is outside the scope of this document.

Please refer to the [documentation](worker-perf) for recommendations on when to use different [slot allocation strategies](slot-alloc-strat) for different workloads.

Briefly, however, follow this advice:

> Scenarios with tasks that have variable, or very high, per-task resource
> needs should rely on fixed-size suppliers and manual tuning rather than
> resource-based suppliers.

[tc-openmetrics]: https://docs.temporal.io/cloud/metrics/openmetrics
[worker-perf]: https://docs.temporal.io/develop/worker-performance
[slot-alloc-strat]: https://docs.temporal.io/develop/worker-performance#choosing-slot-supplier-types

## HPA strengths

Because HPA uses a single OpenMetrics scrape to gather all series for the namespace in a single HTTP request, the HPA approach scales independently of namespace count. The single HTTP request for OpenMetrics more efficient than KEDA's Temporal API-based approach, and will not run into Temporal API rate limiting problems (see section below on [KEDA limitations](#keda-limitations)).

> HPA + prometheus adapter configured to look at both slot util and backlog provides fast scale-up via slot util and a backlog-driven backstop to prevent overly reactive replica count adjustment.

Author: nit: It feels like this sentence is in "note taking" style. Can we say "utilization" instead of "util"? And "can be configured" instead of "configured"? ie:
> HPA + prometheus adapter can be configured to combine slot utilization (or any other other version-labelled metrics) with backlog count. Slot utilization can be used to prevent overly reactive scale-down when backlog is zero but the workers are well-utilized and replica count is right-sized for the workload.

Reviewer: done.

## HPA limitations

This section describes two known limitations for HPA + prometheus adapter.

Temporal Cloud's OpenMetrics endpoint may sometimes return the same embedded timestamps on repeated scrapes for each series across the account simultaneously — backlog series, action counts, error counts, every queue, every namespace. This delay in returning fresh metrics data can impact the speed to which HPA + prometheus adapter scales out or in the replica count for a worker deployment version. This means that HPA + prometheus adapter may not be a good solution if your workload cannot tolerate occasional multi-minute scaling pauses.

> **Warning**: There is an [up to 3 minute potential delay][om-delay] before exported metrics are available in the Temporal Cloud OpenMetrics endpoint for new task queues.

> > **Note**: This is why `metricsRelistInterval: 5m` is the recommended setting: the discovery window must comfortably exceed the longest expected delay so the metric does not deregister, otherwise re-registration waits up to one more relist cycle after delivery resumes.

Author: @dustin-temporal is this correct? / is this the correct way to say it? I know we've discussed in slack, but just want to triple check here also.

  maxReplicas: 30
  metrics:
    - type: External
      external:
        metric:
          name: temporal_slot_utilization
          selector:
            matchLabels:
              worker_type: "ActivityWorker"
        target:
          type: AverageValue
          value: "0.75"

    - type: External
      external:
        metric:
          name: temporal_cloud_v1_approximate_backlog_count
          selector:
            matchLabels:
              temporal_task_queue: "default_helloworld"
              task_type: "Activity"
        target:
          type: AverageValue
          averageValue: "1"
  behavior:
    scaleUp:
      stabilizationWindowSeconds: 30
      policies:
        - type: Percent
          value: 10
          periodSeconds: 10
      selectPolicy: Max

    scaleDown:
      stabilizationWindowSeconds: 120
      policies:
      - type: Percent
        value: 10
        periodSeconds: 10
      selectPolicy: Max
```

## KEDA strengths

KEDA's Temporal scaler calls `DescribeTaskQueue(stats=true)` (or `DescribeWorkerDeploymentVersion`), which loads the queue synchronously and returns the backlog directly. This allows KEDA to scale Temporal workers from zero.

> ## KEDA limitations

Author: also, IMO we should mention here the fact that KEDA Temporal Scaler uses backlog count _only_, and TWC does not currently template per-version metrics queries in other KEDA scalers.

Reviewer: added.
